lib/service/awsoidc.go

```diff
@@ -0,0 +1,245 @@
+package service
+
+import (
+	"context"
+	"errors"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/utils/retryutils"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/auth/integration/integrationv1"
+	"github.com/gravitational/teleport/lib/automaticupgrades"
+	"github.com/gravitational/teleport/lib/integrations/awsoidc"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/utils/interval"
+
+	ststypes "github.com/aws/aws-sdk-go-v2/service/sts/types"
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"golang.org/x/time/rate"
+)
+
+const (
+	// updateDeployAgentsInterval specifies how frequently to check for available updates.
+	updateDeployAgentsInterval = time.Minute * 30
+
+	// updateDeployAgentsRateLimit specifies the time between updates across AWS regions.
+	updateDeployAgentsRateLimit = time.Second * 30
+)
+
+func (process *TeleportProcess) periodUpdateDeployServiceAgents() error {
+	if !process.Config.Auth.Enabled {
+		return nil
+	}
+
+	// start process only after teleport process has started
+	if _, err := process.WaitForEvent(process.GracefulExitContext(), TeleportReadyEvent); err != nil {
+		return nil
+	}
+	process.log.Infof("The new service has started successfully. Checking for deploy service updates every %v.", updateDeployAgentsInterval)
+
+	// Acquire the semaphore before attempting to update the deploy service agents.
+	// This task should only run on a single instance at a time.
+	lock, err := services.AcquireSemaphoreWithRetry(process.GracefulExitContext(),
```

Comment:
Hmm not sure I understand the purpose of this semaphore: do we acquire the lock before starting the service? Shouldn't we acquire it only when actually running the upgrade? Cause the semaphore will expire after updateDeployAgentsInterval and the 2nd process will start anyway?

Reply:
It seems like the semaphore continues to renew as long as the process continues to run. So the 2nd process will only be able to acquire the lease after the 1st process is terminated. So there is only on instance of this task running at a time.

I'll read through the code again and make necessary changes. I'm not entirely confident I'm using the semaphores correctly here.

Reply:
I've made some changes to how I was using the semaphore locks. Let me know if that looks correct.